apps/sms/views/shared/js/activity_handler.js

```diff
@@ -21,63 +24,46 @@ const ActivityDataType = {
 };
 
 var ActivityHandler = {
-  // Will hold current activity object
-  _activity: null,
-
   init: function() {
     if (!window.navigator.mozSetMessageHandler) {
       return;
     }
 
-    // A mapping of MozActivity names to their associated event handler
-    window.navigator.mozSetMessageHandler('activity',
-      this._onActivity.bind(this, {
-        'new': this._onNewActivity,
-        'share': this._onShareActivity
-      })
-    );
-
-    // We don't want to register these system handlers when app is run as
-    // inline activity
-    if (!Navigation.getPanelName().startsWith('activity')) {
-      window.navigator.mozSetMessageHandler('sms-received',
-        this.onSmsReceived.bind(this));
-
-      window.navigator.mozSetMessageHandler('notification',
-        this.onNotification.bind(this));
-    }
-  },
-
-  isInActivity: function isInActivity() {
-    return !!this._activity;
-  },
+    if (ActivityShim.hasPendingRequest()) {
+      // Unique identifier used to have only 1-to-1 connection between client
+      // and corresponding service.
+      var appInstanceId = Date.now();
 
-  setActivity: function setActivity(value) {
-    if (!value) {
-      throw new Error('Activity should be defined!');
-    }
-    this._activity = value;
-  },
+      ActivityClient.init(appInstanceId);
 
-  // The Messaging application's global Activity handler. Delegates to specific
-  // handler based on the Activity name.
-  _onActivity: function activityHandler(handlers, activity) {
-    var name = activity.source.name;
-    var handler = handlers[name];
+      ActivityClient.on(
+        'new-activity-request', this._onNewActivity.bind(this)
+      );
 
-    if (typeof handler === 'function') {
-      this.setActivity(activity);
+      ActivityClient.on(
+        'share-activity-request', this._onShareActivity.bind(this)
+      );
 
-      handler.call(this, activity);
+      // Currently both client and service for activity reside in the same
+      // context, so we spin up both in the same context/window.
+      ActivityShim.init(appInstanceId);
```

Comment:
> I wonder if you'll be able to pass the instanceId to the shim once it lives in another context ?

I've explained one of the ways in https://github.com/mozilla-b2g/gaia/pull/30144#issuecomment-110022887, and we won't need it when we live in the Shared or Service Worker